liquidity/threshold_rule.go

```diff
@@ -59,3 +62,115 @@ func (r *ThresholdRule) validate() error {
+		minimumAmount: btcutil.Amount(
+			uint64(balances.capacity) * uint64(minOutbound) / 100,
+		),
+	}
+
+	switch {
+	// If we have too little inbound and too little outbound, we cannot
+	// do anything to help our liquidity situation. This will happen in the
+	// case where we have a lot of pending htlcs on our channels.
+	case inbound.hasDeficit() && outbound.hasDeficit():
+		return 0, nil
+
+	// If we have too little inbound, but not too little outbound, it is
+	// possible that a loop out will improve our inbound liquidity.
+	case inbound.hasDeficit() && outbound.hasSurplus():
+		amt := calculateSwapAmount(balances.capacity, inbound, outbound)
+		out := swap.TypeOut
+		return amt, &out
+
+	// If we have enough inbound, and too little outbound, it is possible
+	// that we can loop in to improve our outbound liquidity.
+	case inbound.hasSurplus() && outbound.hasDeficit():
+		amt := calculateSwapAmount(balances.capacity, outbound, inbound)
+		in := swap.TypeIn
+		return amt, &in
+
+	// If we have enough inbound and enough outbound, we do not need to
+	// take any actions at present.
+	default:
+		return 0, nil
+	}
+}
+
+// balanceRequirement describes liquidity in one direction.
+type balanceRequirement struct {
+	// currentAmount is the amount of liquidity have have in the direction.
+	currentAmount btcutil.Amount
+
+	// minimumAmount is the minimum amount of liquidity that we require in
+	// this direction.
+	minimumAmount btcutil.Amount
+}
+
+func (i balanceRequirement) hasSurplus() bool {
+	return i.currentAmount > i.minimumAmount
+}
+
+func (i balanceRequirement) hasDeficit() bool {
+	return i.currentAmount < i.minimumAmount
+}
+
+// calculateSwapAmount calculates the amount of capacity that we need to shift
+// to improve the balance of our channels, and returns a reason which provides
+// additional context. This function is used in the case where we have surplus
+// liquidity in one direction, and deficit in another. This is the case in which
+// we recommend swaps, if we have deficits on both sides, we cannot swap without
+// further unbalancing, and if we have surplus on both sides, we do not need to.
+//
+// This function calculates the portion of our total capacity we should shift
+// from the surplus side to deficit side without unbalancing the surplus side.
+// This is important, because we do not want to recommend a swap in one
+// direction that just results in our needing to produce a swap in the other
+// direction.
+func calculateSwapAmount(capacity btcutil.Amount, deficitSide,
+	surplusSide balanceRequirement) btcutil.Amount {
+
+	// Get the maximum shift allowed on our surplus side that is possible
+	// before it dips beneath its minimum.
+	available := surplusSide.currentAmount - surplusSide.minimumAmount
+
+	// Get the midpoint between our two minimum amounts, expressed from
+	// the perspective of the side that we currently have a deficit of
+	// liquidity.
+	midpoint := (deficitSide.minimumAmount +
```

Comment:
We only do loop out now, but the reason that we still have a minimum inbound is to properly calculate the midpoint? Alternatively the mid point could be set directly, `ideal_balance`.

Reply:
I think that specifying thresholds allows us to be a bit more expressive as to when to suggest a swap? 

Take the simple case of a channel with 100 capacity, with thresholds one could set the following, both with a midpoint of 50:
A: `MinIncoming=40` `MinOutgoing=40` 
B: `MinIncoming=10` `MinOutgoing=10`

However, they have quite different meanings, A wants to swap whenever incoming drops below 40%, whereas B is happy to wait till they get all the way down to 10%. 

We also manage to capture the idea of a "sends reserve" with a `minOutbound` which (I think) is lost when we set a single value?